### PR TITLE
[speechlm2] SALMAutomodel: THD (packed sequence) and context parallel support

### PR DESCRIPTION
**Pull request by pzelasko**
> [!IMPORTANT]  
> The `Update branch` button must only be pressed in very rare occassions.
> An outdated branch is never blocking the merge of a PR.
> Please reach out to the automation team before pressing that button.

# What does this PR do ?

Adds opt-in THD packed-sequence training and full context-parallel support to `SALMAutomodel`, with a fit-start validator that catches known-bad configs.

**Collection**: speechlm2

# Changelog

- THD packed sequences (model.packed_sequences: true) — concatenates per-utterance text + audio embeddings into a single flat [T_total, H] sequence with cu_seqlens instead of padding to [B, T_max, H]. Packed sequences still contain _some_ padding for CP. New `parts/packed_sequences.py` + `parts/cp_helpers.get_cp_mesh`.
- Context parallelism (`cp_size > 1`) — sequence-shard hybrid Mamba/attention LLMs across GPUs. Requires TransformerEngine. New `parts/cp_helpers.shard_bshd_for_cp` + `encode_audio_with_cp_distribution`. Recommended pairing is THD + CP. `prepare_inputs` derives the CP mesh once, distributes audio encoding, and (for the BSHD branch) inserts the TE `DualChunkSwap` shard before the TP-truncation fallback.
- Config validator — `parts/parallel.validate_parallelism_compatibility` runs at `on_fit_start` and raises errors for three known-bad combos: BSHD+CP>1 (NaN at step 2 from pad-K/V leak), THD+non-TE attention (Automodel's SDPA THD path is not ready), and THD+TE without `NVTE_FUSED_ATTN=0` (cuDNN backward gradient amp on Blackwell sm_120 — hard error on sm_120, warn elsewhere).
- Docs — `training_and_scaling.rst` and `configs.rst` document the new packed_sequences flag, the CP knob, the BSHD+CP-not-supported warning, and the `NVTE_FUSED_ATTN=0` workaround.

# Usage

- You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"

**Pre checks**:

- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:

- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed.
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Comment by pzelasko** [comment]
/ok to test [9ba5eec](https://github.com/NVIDIA-NeMo/NeMo/pull/15679/commits/9ba5eec3902f83723b9d50b7e0b73d1dbbdab0d5)

**Comment by pzelasko** [comment]
/ok to test [73862e5](https://github.com/NVIDIA-NeMo/NeMo/pull/15679/commits/73862e5fdc14569bdab9bebf3b4c53703e033f03)

**Comment by pzelasko** [comment]
/ok to test [d15c316](https://github.com/NVIDIA-NeMo/NeMo/pull/15679/commits/d15c316ac1c488a250ec614432a7800dd9de60ca)

File: nemo/collections/speechlm2/parts/cp_helpers.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+
+    local_stack = torch.zeros(per_rank, max_L, H, device=device, dtype=local_embs[0].dtype)
+    local_lens = torch.zeros(per_rank, dtype=torch.long, device=device)
+    for i, e in enumerate(local_embs):
+        local_stack[i, : e.shape[0]] = e
+        local_lens[i] = e.shape[0]
+
+    gathered_stack = [torch.zeros_like(local_stack) for _ in range(cp_size)]
+    gathered_lens = [torch.zeros_like(local_lens) for _ in range(cp_size)]
+    dist.all_gather(gathered_stack, local_stack, group=cp_mesh.get_group())
```
Comment (KunalDhawan) [inline nemo/collections/speechlm2/parts/cp_helpers.py]: Blocker: looks like `dist.all_gather` isn't autograd-aware, `gathered_stack` comes back with no `grad_fn`. Since `full_embs` is built entirely from `gathered_stack`, the perception module receives no gradient at all when CP is active. But in In the `cp_size==1` path, `encode_audio_with_optional_chunking` returns grad-connected embeddings, so this is also a silent behavioral divergence between CP and non-CP. Is this intended? This would be safe only when `self.perception` is fully frozen, which might not always be the case.

File: nemo/collections/speechlm2/models/salm_automodel.py [hunk 2]
```diff
@@ -318,13 +386,18 @@ def validation_step(self, batch: dict, batch_idx: int):
...
-            forward_outputs = self(inputs["input_embeds"], attention_mask=inputs["attention_mask"])
+            forward_outputs = self(
+                inputs["input_embeds"],
+                attention_mask=inputs["attention_mask"],
+                **inputs.get("llm_kwargs", {}),
+            )
             num_frames = (inputs["target_ids"] != -100).long().sum()
             with loss_parallel():
+                logits = forward_outputs["logits"]
                 loss = (
```
Comment (KunalDhawan) [inline nemo/collections/speechlm2/models/salm_automodel.py]: Here each rank would hold only a shard of a global sequence's tokens under CP, so `local_CE_sum` / `local_num_frames` would be a per-shard ratio. `on_validation_epoch_end` would then take an unweighted mean() (with sync_dist=True) over ranks with different denominators, which isn't the true `Σsum / Σframes`. Accuracy looks to have the same micro-vs-macro issue, since `preds.eq(refs).float().mean()` is averaged across unequal shards. This wouldn't affect training but biases the val_acc used for checkpoint selection. Should we mirror the non-CP path -  aggregate loss_sum and num_frames separately across the dp_cp group and divide once?

File: nemo/collections/speechlm2/parts/packed_sequences.py [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    # Round each utterance's length up to a multiple of 2*cp_size (TE-CP
+    # interleaves 2 chunks per rank); skip rounding when cp_size == 1. Then
+    # bump the last so the total is divisible by tp_size for sequence
+    # parallelism.
+    if cp_size > 1:
+        cp_mult = 2 * cp_size
+        padded_lens = [((L + cp_mult - 1) // cp_mult) * cp_mult for L in real_lens]
+    else:
+        padded_lens = list(real_lens)
+    if tp_size > 1:
```
Comment (KunalDhawan) [inline nemo/collections/speechlm2/parts/packed_sequences.py:153]: minor nit: Each per-utt length is rounded to a multiple of `2*cp_size`, but the final `tp_size` bump adds `tp_size - rem` only to the last segment. When tp_size isn't a multiple of `2*cp_size`, that bump breaks the last segment's `2*cp_size` alignment, which violates the `thd_get_partitioned_indices` contract. The current test (`test_tp_and_cp_combined`) has cp_size=2, tp_size=8, so this issue is not hit. Maybe add an assert `tp_size % (2*cp_size) == 0 ` here?

**Review by KunalDhawan — CHANGES_REQUESTED** [review]

File: nemo/collections/speechlm2/parts/cp_helpers.py (hunk 1 above)
Comment (pzelasko) [reply nemo/collections/speechlm2/parts/cp_helpers.py]: great catch, thanks

**Comment by pzelasko** [comment]
/ok to test [34fca17](https://github.com/NVIDIA-NeMo/NeMo/pull/15679/commits/34fca1791981570b0b02ae85d00c6d6c4ab0dff6)

**Comment by pzelasko** [comment]
/ok to test [095ce23](https://github.com/NVIDIA-NeMo/NeMo/pull/15679/commits/095ce23e19ebce0f2942e5e2d5e952a90a419a45)

**Review by KunalDhawan — APPROVED** [review]
Great work, thanks @pzelasko!
